electrum_nmc/electrum/names.py

```diff
@@ -132,6 +132,230 @@ def validate_value_length(value: bytes):
+
+        # Check the version byte
+        version_byte = decoded_service_id[V3_ONION_SERVICE_ID_VERSION_OFFSET]
+        if version_byte > 0x03:
+            raise ValueError(f"Warning: this is a v{version_byte} onion service, which is newer than the v3 onion services that Electrum-NMC knows how to validate. Please verify that this is what you really meant to enter.")
+
+        if version_byte < 0x03:
+            raise ValueError(f"Obsolete v{version_byte} onion service")
+
+        # Extract the public key from the decoded service ID
+        public_key = bytearray(decoded_service_id[:ED25519_PUBLIC_KEY_SIZE])
+
+        # Calculate the truncated checksum
+        truncated_checksum = calc_onion_truncated_checksum(public_key)
+
+        # Check if the truncated checksum matches the corresponding bytes in the decoded service ID
+        if truncated_checksum[0] != decoded_service_id[V3_ONION_SERVICE_ID_CHECKSUM_OFFSET] or \
+                truncated_checksum[1] != decoded_service_id[V3_ONION_SERVICE_ID_CHECKSUM_OFFSET + 1]:
+            raise ValueError("Invalid checksum")
+
+    except Exception as e:
+        raise ValueError(f"Invalid onion address: {e}")
+
+def calc_onion_truncated_checksum(public_key:str):
+    hasher = hashlib.sha3_256()
+
+    # Calculate the checksum
+    hasher.update(b".onion checksum")
+    hasher.update(public_key)
+    hasher.update(bytes([0x03]))
+    hash_bytes = bytearray(hasher.digest())
+
+    return hash_bytes[:2]
+
+def validate_ipv4_address(address: str) -> None:
+    try:
+        ipaddress.IPv4Address(address)
+    except ipaddress.AddressValueError as e:
+        raise ValueError(f"Invalid IPv4 address: {e}")
+    except ipaddress.NetmaskValueError as e:
+        raise ValueError(f"Invalid IPv4 netmask: {e}")
+
+def validate_ipv6_address(address: str) -> None:
+    try:
+        ip = ipaddress.IPv6Address(address)
+        if address != ip.compressed:
+            raise ValueError("IPv6 address should be in compressed form")
+    except ipaddress.AddressValueError as e:
+        raise ValueError(f"Invalid IPv6 address: {e}")
+    except ipaddress.NetmaskValueError as e:
+        raise ValueError(f"Invalid IPv6 netmask: {e}")
+
+def validate_zeronet_address(address: str) -> None: # Validate P2PKH Address (Zero Net)
+    try:
+        addrtype, _ = b58_address_to_hash160(address)
+    except Exception as e:
+        raise ValueError(f"Invalid zeronet address: {e}")
+    if addrtype != ZERONET_BASE58_ADDRTYPE:
+        raise ValueError("Invalid address type.")
+
+def validate_i2p_address(address: str) -> None:
+    try:
+        # Check if it ends with '.b32.i2p'
+        if not address.endswith('.b32.i2p'):
+            raise ValueError("Address doesn't end with '.b32.i2p'")
+
+        # remove '.b32.i2p'
+        address = address[:-len('.b32.i2p')] 
+        # Check length
+        if len(address) > 67:
+            raise ValueError("Address length exceeds 67 characters")
```

Comment:
It would be good to have a citation in a comment explaining why 67 is the max. (Is there a spec somewhere that says this?) Also is there a min?

Reply:
Yes, but it doesn't talk about the minimum. https://geti2p.net/en/docs/naming

Reply:
grep `67 characters maximum`

Reply:
The only information is about the Minimum key length (516 bytes) but since we already check `if len(decoded) != BASE32_LENGTH:` checking the Minimum key might not be necessary.